src/main/java/org/opensearch/knn/index/query/ExactSearcher.java

```diff
@@ -60,11 +60,13 @@ public class ExactSearcher {
     public Map<Integer, Float> searchLeaf(final LeafReaderContext leafReaderContext, final ExactSearcherContext exactSearcherContext)
         throws IOException {
         KNNIterator iterator = getKNNIterator(leafReaderContext, exactSearcherContext);
+        if (iterator == null) {
+            return Collections.emptyMap();
+        }
         if (exactSearcherContext.getKnnQuery().getRadius() != null) {
             return doRadialSearch(leafReaderContext, exactSearcherContext, iterator);
         }
-        if (exactSearcherContext.getMatchedDocs() != null
-            && exactSearcherContext.getMatchedDocs().cardinality() <= exactSearcherContext.getK()) {
+        if (exactSearcherContext.getMatchedDocs() != null && exactSearcherContext.getMatchedDocs().cost() <= exactSearcherContext.getK()) {
```

Comment:
the `cost()` not necessary always give the right cost. It depends on the underline implementation of the iterator. So what Implementation of iterator will be present under the hood and how correctly does `cost()` function gives the number of docs.

Reply:
That's a good point. I focused primarily on the one that I passed here (GroupedNestedDocIdSetIterator) to ensure the cost calculation isn't an approximation.

When I reviewed the code, it seems the main difference between scoreAllDocs and searchTopCandidates is whether a heap is used to maintain the top-k results. Would it make sense to always call searchTopCandidates, considering the latency difference might be negligible?

Alternatively, I can introduce a new class that holds the iterator along with the accurate cost value, if that would be preferable.

Reply:
> When I reviewed the code, it seems the main difference between scoreAllDocs and searchTopCandidates is whether a heap is used to maintain the top-k results. Would it make sense to always call searchTopCandidates, considering the latency difference might be negligible?

The difference between both of them in 1 case when heap is used we can have more than k docs and in case no heap, then we don't need heap. The latency difference will not be negligible when done enough number of times. This is the reason why the 2 functions were created because having a heaps add extra latency.

Reply:
> Alternatively, I can introduce a new class that holds the iterator along with the accurate cost value, if that would be preferable.

If you already have the cost of iterator then I would recommend passing it along in this ExactSearcherContext class, to ensure we find the cost of these iterators/bitset only once.